CRDCDH-2594 Add support for auth permission scoping

Alejandro-Vega opened the pull request:
### Overview

In the future we will be adding scoping to permissions, so this is to mostly ignore it for now. Although, I added support for it.

### Change Details (Specifics)

- Added support for future extensions within hasPermissions
- Removed any permission.includes that I could find, only found within navigation and permission panel. Replaced it to clean up the permission and disregard the scoping
  - Currently the BE throws an error if you pass a permission with a scope while setting the PBAC panel, so I have it removing anything other than the entity:action for now. Can be updated if needed
- The types for Notifications can likely be removed, but I think we hold off for now. Perhaps we just type it to just a string instead

### Related Ticket(s)

[CRDCDH-2594](https://tracker.nci.nih.gov/browse/CRDCDH-2594) (Task)
[CRDCDH-2656](https://tracker.nci.nih.gov/browse/CRDCDH-2656) (US)

amattu2 reviewed (commented):
I haven't finished fully reviewing yet, but I wanted to pause here because I think I've confused myself. Does the `hasPermission` function need to be called with scope details (`opts`)? 

My thought was,

1. Component calls hasPermission asking about a certain Entity:Action
2. hasPermission passes request off to action verifier (permission map) 
3. Result propagated to component

In this case, only (2) needs to handle scope (once we implement that). I'm trying to understand when would the component know what scopes are relevant to it? 

Again... I think I confused myself, this doesn't necessarily mean your implementation is incorrect.

Alejandro-Vega commented on src/config/AuthPermissions.ts:
```diff
@@ -213,39 +243,52 @@ export const PERMISSION_MAP = {
+ *   Optional configuration:
+ *   - `data` (`Permissions[E]["dataType"]`): payload for dynamic checks.
+ *   - `onlyKey` (`boolean`): if true, skip all logic except the permission key.
+ *   - `opts` (`PermissionOpts<S>`): scope filters (`scope` + `scopeValues`).
```
Outdated, reverted to old params and documentation.

Alejandro-Vega commented on src/utils/authUtils.ts:
Moved authUtils to profileUtils in [c1fb046](https://github.com/CBIIT/crdc-datahub-ui/pull/691/commits/c1fb0469fddab9be1356bff2f9804fdab37bb189). Thanks!

Alejandro-Vega commented:
> I haven't finished fully reviewing yet, but I wanted to pause here because I think I've confused myself. Does the `hasPermission` function need to be called with scope details (`config`.`opts`)?
> 
> My thought was,
> 
> 1. Component calls hasPermission asking about a certain Entity:Action
> 2. hasPermission passes request off to action verifier (permission map)
> 3. Result propagated to component
> 
> In this case, only (2) needs to handle scope (once we implement that). I'm trying to understand when would the component know what scopes are relevant to it?
> 
> Again... I think I confused myself, this doesn't necessarily mean your implementation is incorrect.

After Slack discussion, I have reworked the implementation to the following:

- Take in the entity and action params within hasPermission, which we already do (Ex. "data_submission" and "create")
- Use that to look through the users permissions and find the matching permission key
- If we pass onlyKey param, then we skip the scope check
- Otherwise, check for scope/values if they have any
- Pass that along to the AuthPermissions action map as a param. Will be handled in the future.
- In the future, we would use this param to refactor the role-based checking and instead use the scope

amattu2 reviewed (approved):
Changes LGTM and seems to work as expected. I'm approving this now, but let's wait to merge until BE has their PBAC refactoring deployed.

amattu2 reviewed (changes requested):
@Alejandro-Vega Ming just confirmed that the PBAC Defaults will contain the scoping already attached to it. This means:

- We can remove the scope-stripping before sending it to the backend
- And I think (??) we can just revert the changes in the PBAC Panel

Alejandro-Vega commented:
> @Alejandro-Vega Ming just confirmed that the PBAC Defaults will contain the scoping already attached to it. This means:
> 
> * We can remove the scope-stripping before sending it to the backend
> * And I think (??) we can just revert the changes in the PBAC Panel

Thanks Alec! I reverted the scope-stripping and and PBAC Panel changes in [b5e8a63](https://github.com/CBIIT/crdc-datahub-ui/pull/691/commits/b5e8a636f928f002e6f4d65446993b585cdc6aad). Also, added additional test in [f767931](https://github.com/CBIIT/crdc-datahub-ui/pull/691/commits/f767931762166c1eab61463ba97aba0f0321b5ef) to ensure the checkboxes will still work with the scoping included in the defaults and user perms.

amattu2 reviewed (approved):
LGTM. Thanks for working through all of the back-and-forth discussions/revisions.

The BE changes aren't deployed yet, but this should be compatible AS-IS. Merging.
